bindings/rust/s2n-tls-tokio/tests/shutdown.rs

```diff
@@ -76,13 +79,65 @@ async fn shutdown_after_split() -> Result<(), Box<dyn std::error::Error>> {
 
     let mut received = [0; 1];
 
-    // ignore any shutdown errors in case things go wrong
-    let _ = tokio::try_join!(
+    // All tasks must cleanly exit. try_join will return as soon as an error
+    // occurs, so if the result is any error then the test has failed.
+    tokio::try_join!(
         server.shutdown(),
         client_reader.read(&mut received),
         write_until_shutdown(&mut client_writer),
-    );
+    )?;
+    Ok(())
+}
+
+/// Reading and writing handles should both respond to a peers "close notify"
+/// appropriately. TLS1.3 connections have "half close behavior". The read
+/// handle should immediately exit, but the write handle can continue to write
+/// until explicitly shutdown. After both client handles have shutdown, the
+/// server should cleanly exit.
+#[tokio::test]
+async fn shutdown_after_halfclose_split() -> Result<(), Box<dyn std::error::Error>> {
+    let (server_stream, client_stream) = common::get_streams().await?;
+
+    let client = TlsConnector::new(common::client_config()?.build()?);
+    let server = TlsAcceptor::new(common::server_config()?.build()?);
 
+    let (client, mut server) =
+        common::run_negotiate(&client, client_stream, &server, server_stream).await?;
+
+    let (mut client_reader, mut client_writer) = tokio::io::split(client);
+
+    let close_notify_recvd = Arc::new(tokio::sync::Notify::new());
+    let close_notify_recvd_clone = close_notify_recvd.clone();
+
+    let mut received = [0; 1];
+
+    // all tasks must complete, and must complete successfully
+    // the client tasks will panic if an error is encountered, so those don't
+    // need to be checked.
+    let (server, _, _) = tokio::join!(
+        server.shutdown(),
+        async {
+            let bytes_read = client_reader.read(&mut received).await.unwrap();
+            // 0 bytes read indicate that we returned because of close notify
+            assert_eq!(bytes_read, 0);
+            // signal the writer task that close notify received
+            close_notify_recvd.notify_one();
+        },
+        async {
+            // wait for the connection to receive "close notify" from peer
+            close_notify_recvd_clone.notified().await;
+            // confirm that we can write even after receiving the shutdown from
+            // the server
+            client_writer
+                .write_all("random bytes".as_bytes())
```

Comment:
Shouldn't we assert on the server side that these bytes were received?

Reply:
Well, ideally yes. But that's not really an option with our current implementation
Behavior after `s2n_shutdown`
```
 * Once `s2n_shutdown` is complete:
 * * The s2n_connection handle cannot be used for reading for writing.
 * * The underlying transport can be closed. Most likely via `shutdown()` or `close()`.
 * * The s2n_connection handle can be freed via s2n_connection_free() or reused via s2n_connection_wipe()
```

Behavior after `s2n_shutdown_send`
```
 * Once `s2n_shutdown_send()` is complete:
 * * The s2n_connection handle CANNOT be used for writing.
 * * The s2n_connection handle CAN be used for reading.
 * * The write side of the underlying transport can be closed. Most likely via `shutdown()`.
```

Out tokio implementation uses `s2n_shutdown`, so we can't read after having sent the shutdown. There might be a way around that, but it would not be straightforward with our current tokio bindings.

The alternate way to look at this is that we are just asserting half-close functionality for the client, not for the server, because  as far as I understand we don't currently expose half-close functionality for self-initiated shutdowns in `s2n-tls-tokio`.